File: clients/scala/src/main/scala/ai_cup_22/debugging/DebugState.scala
package ai_cup_22.debugging

import ai_cup_22.util.StreamUtil

/**
 * Renderer's state
 *
 * @param pressedKeys Pressed keys
 * @param cursorWorldPosition Cursor's position in game coordinates
 * @param lockedUnit Id of unit which is followed by the camera, or None
 * @param camera Current camera state
 */
case class DebugState(pressedKeys: Seq[String], cursorWorldPosition: ai_cup_22.model.Vec2, lockedUnit: Option[Int], camera: ai_cup_22.debugging.Camera) {
    /**
     * Write DebugState to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeInt(stream, pressedKeys.length)
        pressedKeys.foreach { value =>
            StreamUtil.writeString(stream, value)
        }
        cursorWorldPosition.writeTo(stream)
        lockedUnit match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
        camera.writeTo(stream)
    }

    /**
     * Get string representation of DebugState
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("DebugState { ")
        stringBuilder.append("pressedKeys: ")
        stringBuilder.append(pressedKeys)
        stringBuilder.append(", ")
        stringBuilder.append("cursorWorldPosition: ")
        stringBuilder.append(cursorWorldPosition)
        stringBuilder.append(", ")
        stringBuilder.append("lockedUnit: ")
        stringBuilder.append(lockedUnit)
        stringBuilder.append(", ")
        stringBuilder.append("camera: ")
        stringBuilder.append(camera)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object DebugState {
    /**
     * Read DebugState from input stream
     */
    def readFrom(stream: java.io.InputStream): DebugState = DebugState(
        (0 until StreamUtil.readInt(stream)).map { _ =>
            StreamUtil.readString(stream)
        },
        ai_cup_22.model.Vec2.readFrom(stream),
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None,
        ai_cup_22.debugging.Camera.readFrom(stream)
    )
}

File: clients/cpp_variant/debugging/DebugState.cpp
#include "DebugState.hpp"

namespace debugging {

DebugState::DebugState(std::vector<std::string> pressedKeys, model::Vec2 cursorWorldPosition, std::optional<int> lockedUnit, debugging::Camera camera) : pressedKeys(pressedKeys), cursorWorldPosition(cursorWorldPosition), lockedUnit(lockedUnit), camera(camera) { }

// Read DebugState from input stream
DebugState DebugState::readFrom(InputStream& stream) {
    std::vector<std::string> pressedKeys = std::vector<std::string>();
    size_t pressedKeysSize = stream.readInt();
    pressedKeys.reserve(pressedKeysSize);
    for (size_t pressedKeysIndex = 0; pressedKeysIndex < pressedKeysSize; pressedKeysIndex++) {
        std::string pressedKeysElement = stream.readString();
        pressedKeys.emplace_back(pressedKeysElement);
    }
    model::Vec2 cursorWorldPosition = model::Vec2::readFrom(stream);
    std::optional<int> lockedUnit = std::optional<int>();
    if (stream.readBool()) {
        int lockedUnitValue = stream.readInt();
        lockedUnit.emplace(lockedUnitValue);
    }
    debugging::Camera camera = debugging::Camera::readFrom(stream);
    return DebugState(pressedKeys, cursorWorldPosition, lockedUnit, camera);
}

// Write DebugState to output stream
void DebugState::writeTo(OutputStream& stream) const {
    stream.write((int)(pressedKeys.size()));
    for (const std::string& pressedKeysElement : pressedKeys) {
        stream.write(pressedKeysElement);
    }
    cursorWorldPosition.writeTo(stream);
    if (lockedUnit) {
        stream.write(true);
        const int& lockedUnitValue = *lockedUnit;
        stream.write(lockedUnitValue);
    } else {
        stream.write(false);
    }
    camera.writeTo(stream);
}

// Get string representation of DebugState
std::string DebugState::toString() const {
    std::stringstream ss;
    ss << "DebugState { ";
    ss << "pressedKeys: ";
    ss << "[ ";
    for (size_t pressedKeysIndex = 0; pressedKeysIndex < pressedKeys.size(); pressedKeysIndex++) {
        const std::string& pressedKeysElement = pressedKeys[pressedKeysIndex];
        if (pressedKeysIndex != 0) {
            ss << ", ";
        }
        ss << '"' << pressedKeysElement << '"';
    }
    ss << " ]";
    ss << ", ";
    ss << "cursorWorldPosition: ";
    ss << cursorWorldPosition.toString();
    ss << ", ";
    ss << "lockedUnit: ";
    if (lockedUnit) {
        const int& lockedUnitValue = *lockedUnit;
        ss << lockedUnitValue;
    } else {
        ss << "none";
    }
    ss << ", ";
    ss << "camera: ";
    ss << camera.toString();
    ss << " }";
    return ss.str();
}

}